File: LeetCode/1334. Find the City With the Smallest Number of Neighbors at a Threshold Distance/solution.py
# class Solution:
#     def findTheCity(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
#         graph = defaultdict(list)
#         for a, b, w in edges:
#             graph[a].append((b,w))
#             graph[b].append((a,w))
#         def func(start):
#             dist = [float('inf')]*n
#             dist[start] = 0
#             q = [(0,start)]
#             heapq.heapify(q)
#             res = 0
#             while q:
#                 dcur, cur = heapq.heappop(q)
#                 if dist[cur] > distanceThreshold:
#                     break
#                 if dcur > dist[cur]:
#                     continue
#                 res += 1
#                 for nxt, d in graph[cur]:
#                     if dist[cur]+d < dist[nxt]:
#                         dist[nxt] = dist[cur]+d
#                         heapq.heappush(q,(dist[nxt],nxt))
#             return res, -start
#         # print([func(i) for i in range(n)])
#         return min(range(n), key=func)
            
            
class Solution:
#     def findTheCity(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
#         dp = defaultdict(int)
#         for i in range(n):
#             for j in range(n):
#                 if i != j:
#                     dp[-1,i,j] = float('inf')
#         for a, b, w in edges:
#             dp[-1,a,b] = w 
#             dp[-1,b,a] = w 
#         for k in range(n):
#             for i in range(n):
#                 for j in range(n):
#                     dp[k,i,j] = min(dp[k-1,i,j],dp[k-1,i,k]+dp[k-1,k,j])
#         return min(range(n), key = lambda x: (sum(dp[n-1,x,y] <= distanceThreshold for y in range(n)), -x))
        
        
    def findTheCity(self, n, edges, maxd):
        dis = [[float('inf')] * n for _ in range(n)]
        for i, j, w in edges:
            dis[i][j] = dis[j][i] = w
        for i in range(n):
            dis[i][i] = 0
        for k in range(n):
            for i in range(n):
                for j in range(n):
                    dis[i][j] = min(dis[i][j], dis[i][k] + dis[k][j])
        res = {sum(d <= maxd for d in dis[i]): i for i in range(n)}
        return res[min(res)]
                
                

File: LeetCode/1334. Find the City With the Smallest Number of Neighbors at a Threshold Distance/solution.cpp
const int inf = 0x3f3f3f3f;

class Solution {
public:
    int findTheCity(int n, vector<vector<int>>& edges, int distanceThreshold) {
        vector<vector<int>> dis(n, vector<int>(n,inf));
        for (int i = 0; i < n; i++)
            dis[i][i] = 0;
        for (auto e : edges){
            dis[e[0]][e[1]] = e[2];
            dis[e[1]][e[0]] = e[2];
        }
        for (int i = 0; i < n; i++){
            for (int u = 0; u < n; u++){
                for (int v = 0; v < n; v++){
                    dis[u][v] = min(dis[u][v], dis[u][i]+dis[i][v]);
                }
            }
        }
        int idx = -1, cnt = inf;
        for (int u = 0; u < n; u++){
            int cc = 0;
            for (int v = 0; v < n; v++)
                if (dis[u][v] <= distanceThreshold)
                    cc++;
            if (cc <= cnt){
                cnt = cc;
                idx = u;
            }
        }
        return idx;
    }
};